test/distributed/fsdp/test_fsdp_use_orig_params.py

```diff
@@ -0,0 +1,875 @@
+        # combination with the parameter group construction, ensures different
+        # hyperparameter settings within one `FlatParameter`
+        fsdp_kwargs = {
+            "auto_wrap_policy": functools.partial(
+                transformer_auto_wrap_policy,
+                transformer_layer_cls={
+                    TransformerEncoderLayer,
+                    TransformerDecoderLayer,
+                },
+            ),
+            "use_orig_params": True,
+            "sharding_strategy": sharding_strategy,
+            "backward_prefetch": backward_prefetch,
+            "cpu_offload": cpu_offload,
+        }
+        model = TransformerWithSharedParams.init(
+            self.process_group,
+            FSDPInitMode.NO_FSDP,
+            cuda_init_mode,
+            deterministic=True,
+        )
+        if init_optim_before_wrap:
+            fsdp_optim = self._get_optim(model, optim_class, multi_tensor)
+            fsdp_model = FSDP(model, self.process_group, **fsdp_kwargs)
+        else:
+            fsdp_model = FSDP(model, self.process_group, **fsdp_kwargs)
+            fsdp_optim = self._get_optim(fsdp_model, optim_class, multi_tensor)
+        if (
+            cuda_init_mode == CUDAInitMode.CUDA_AFTER
+            and not fsdp_model.cpu_offload.offload_params
+        ):
+            fsdp_model = fsdp_model.cuda()
+        return fsdp_model, fsdp_optim
+
+    def _check_train_parity(
+        self,
+        ddp_model: DDP,
+        ddp_optim: torch.optim.Optimizer,
+        fsdp_model: FSDP,
+        fsdp_optim: torch.optim.Optimizer,
+        set_to_none: bool,
+        num_iters: int = 10,
+    ):
+        """Checks training parity between DDP and FSDP."""
+        device = torch.device("cuda")
+        for i in range(num_iters):
+            iter_losses = []
+            for model, optim in ((ddp_model, ddp_optim), (fsdp_model, fsdp_optim)):
+                module = model.module
+                # Test two different `zero_grad()` timings
+                if i % 2 == 0:
+                    optim.zero_grad(set_to_none=set_to_none)  # pre-forward
+                inp = module.get_input(device)
+                output = model(*inp)
+                loss = module.get_loss(inp, output).to(device)
+                iter_losses.append(loss)
+                if i % 2 == 1:
+                    optim.zero_grad(set_to_none=set_to_none)  # pre-backward
+                module.run_backward(loss)
+                # Perform the DDP optimizer step on CPU to match FSDP if needed
+                if model is ddp_model and fsdp_model.cpu_offload.offload_params:
+                    model.to(torch.device("cpu"))
+                optim.step()
+                if model is ddp_model and fsdp_model.cpu_offload.offload_params:
+                    model.to(torch.device("cuda"))
+            torch.testing.assert_close(iter_losses[0], iter_losses[1])
+            iter_losses.clear()
+        atol = 1e-5 if self.world_size <= 2 else 1e-4
```

Comment:
hmm, why do we need this?

Reply:
Interesting. I removed the explicit `atol` and `rtol`, and the tests seem fine with world size of 8. I am not sure what happened when I first prototyped (unless there is some non-negligible non-determinism with the numerics).

Reply:
Maybe try several times to ensure there is no flakiness?